docs/src/architecture/domain-model.md

```diff
@@ -0,0 +1,214 @@
+
+<!--
+* TODO: add somehow "Such a **device** is most often a specialized hardware that has been ruggedized to operate in harsh electrical, thermic and mechanical industrial environments."
+-->
+* The **Domain Application** is a program, that contains the domain specific process logic.
+  * On a **PLC** it's a _control application_, created by an _automation engineer_ in a domain specific graphical programming language (like "Ladder Diagram" or "Function block diagram")
+  * Or on a **SoC-based** or **Microcontroller-based** system it's an _application program_, created by an _embedded software engineer_ usually in C/C++
+* The **OS / Libs / Runtime** provide basic functionality to the **Domain Application**
+  <!-- TODO: add somehow "used to abtracts the hardware. But: on a microcontroler usually less abtraction, more hw dep on the domain app, and even no OS" -->
+* **Sensors** and **Actuators** enable the **device** to observe and control physical behaviour on the shopfloor or device's environment.
+  * can be integrated in the **device's** hardware,
+    or connected to the **device** via some **Fieldbus** (e.g. Modbus, CANopen, PROFINET, EtherCAT, ...) or
+    some **Local Interface** as USB, UART, SPI, I2C, ...
+  * can be simple peripherals as a light barrier, or a sensor for temperature or pressure;
+    or complex equipments as robotic arms or even other **devices**
+* **Inputs / Outputs** are the communication channels between the **Domain Application** and **Sensors** and **Actuators**
+  * drivers (as part of the **OS / Libs / Runtime** and/or the **Domain Application**) do expose all data from
+    **Sensors** and **Actuators** to the **device** as inputs or outputs
+  * also the **Domain Application** can expose data as input or output (e.g. own _signals_ or _states_)
+
+## IoT Cloud Domain
+
+Cloud end-points could be IoT Cloud Platforms, as e.g. [Cumulocity](https://www.softwareag.cloud/site/product/cumulocity-iot.html), [Azure](https://azure.microsoft.com) or [AWS](https://aws.amazon.com); or end-points for specific IoT tasks, as e.g. a time-series database to collect measurements.
+
+* Each **device** is represented in the IoT cloud by an individual **device twin**
+* All information / data or functionality the device provides to the cloud appears in context of it's **device twin**
+* The focus of **thin-edge** is on two IoT Cloud aspects:
+  - 1) **Telemetry Data Handling**
+  - 2) **Device Management**
+
+### Telemetry Data Handling
+
+**Telemetry data** are the measurements, events and alarms collected or raised by sensors, actuators or domain applications.
+
+  * **Measurement**, is a single value or set of values
+    * could be be a mix of values produced by one or more sensors and values calculated by the device's domain application
+    * values could be a mix of numbers, strings or booleans
+    * has one timestamp
+    * releates to one **Metric**
+  * **Metric**, is a time-series of measurements
```

Comment:
I am a bit irretated by introducing "metric" here.
I would move that to some seperate chapter.
IMHO "metric is a special kind of measurement, which represents technical properties of of the device itself like CPU, Memory, etc."
BTW: That would not even be true if someone uses collectd with Modbus collector.

Reply:
@vaillant, I suggested this term to @cstoidner, as this is for me a global term that encompasses values read on the device itself as well as measurements read from sensors.

Reply:
Though do we need both `metrics` and `measurement` here? The first sentence of the `Telemetry data` section does not mention metrics. We could just use `measurement` to simplify the mapping between different domain models (e.g. Cumulocity and tedge).

The different between a measure(ment) and metric is largely academic, and you could still transport a metric as a measurement depending on how you want to model it. And I believe the users are not interested in the difference, and would favour simplicity (less choices)

https://www.simplekpi.com/Blog/metrics-and-meassures-a-definitive-guide

Reply:
The more I investigate about metrics and measurements I came now also to the clue, that metric is not the right thing here.

However it makes sense to distinguish for a measurement between:
- sampled values
   - a measurement's set of values, sampled at one point in time
- meta information
  - a measurement's type, number of values, units per values, ...more?

The latter one should possible to be exist without the first one (especially in start phase when not yet any value was sampled). So I am about to remove "metrics" from the model and adapt the definition of measurements.

Reply:
I'm okay to use a single term. However, there is one point that matters to my academic background: time. The goal is not to have a single measurement but to observe a time-varying value taking samples with a specific frequency. I want to distinguish the sample from the whole series - at least at the conceptual level. This is because we are observing a continuous system that we can increase/decrease the sampling rate and interpolate missing samples or group close measurements as we do for collectd. Note that this is what makes the difference between events and measurements. If you missed an event, you cannot guess it from two events happening a bit sooner and later. It make also no sense to observe events at a different rate.

So, yes, it's important to simplify things (notably, I agree that the difference made in the article linked by @reubenmiller is subtle) but not to the point where we start to confuse key points (up to the point to blur the difference between measurements and events).

Reply:
Adopted in new PR, see https://github.com/thin-edge/thin-edge.io/pull/1554#issuecomment-1307364723